src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+                    // of the compilation. Only the MVID, name and version contribute here hence the file path
+                    // is deliberately omitted here.
+                    var peReader = moduleMetadata.GetMetadataReader();
+                    if (peReader.IsAssembly)
+                    {
+                        var assemblyDef = peReader.GetAssemblyDefinition();
+                        writer.Write("name", peReader.GetString(assemblyDef.Name));
+                        WriteVersion(writer, "version", assemblyDef.Version);
+                        WriteByteArrayValue(writer, "publicKey", peReader.GetBlobBytes(assemblyDef.PublicKey).AsSpan());
+                    }
+                    else
+                    {
+                        var moduleDef = peReader.GetModuleDefinition();
+                        writer.Write("name", peReader.GetString(moduleDef.Name));
+                    }
+
+                    writer.Write("mvid", GetGuidValue(moduleMetadata.GetModuleVersionId()));
+                }
+            }
+            else if (reference is CompilationReference compilationReference)
+            {
+                var compilation = compilationReference.Compilation;
+                var builder = compilation.Options.CreateDeterministicKeyBuilder();
+                builder.WriteCompilation(
+                    writer,
+                    compilation.Options,
+                    compilation.SyntaxTrees.SelectAsArray(static x => SyntaxTreeKey.Create(x)),
+                    compilation.References.AsImmutable(),
+                    compilation.Assembly.Identity.PublicKey,
+                    pathMap,
+                    deterministicKeyOptions,
+                    cancellationToken);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(reference);
+            }
+
+            writer.WriteObjectEnd();
+
+            static void writeMetadataReferenceProperties(JsonWriter writer, MetadataReferenceProperties properties)
+            {
+                writer.WriteObjectStart();
+                writer.Write("kind", properties.Kind);
+                writer.Write("embedInteropTypes", properties.EmbedInteropTypes);
+                writer.WriteKey("aliases");
+                writer.WriteArrayStart();
+                foreach (var alias in properties.Aliases)
+                {
+                    writer.Write(alias);
+                }
+                writer.WriteArrayEnd();
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteEmitOptions(
+            JsonWriter writer,
+            EmitOptions? options,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions deterministicKeyOptions)
+        {
+            writer.WriteObjectStart();
+            if (options is not null)
+            {
+                writer.Write("emitMetadataOnly", options.EmitMetadataOnly);
```

Comment:
As we add new options (or more generally extend the compilation object model), is there anything we can do to remind us to update this code?
An entry mentioning determinism key may be useful in the compiler test plan. Maybe comments in the relevant types might help too?

Reply:
The unit tests are designed to help guard against this because they effectively count the number of properties on the critical types. If a new property is added then the tests will fail and the test has a comment about making sure to update this representation. 

If you have suggestions on how to make these tests more robust I'm def open to them.

Reply:
See the `VerifyUpToDate` tests

Reply:
Thanks. `VerifyUpToDate` seems a good approach